core/src/main/java/org/jdbi/v3/core/qualifier/Qualifiers.java

```diff
@@ -36,40 +38,87 @@
+     * which searches for annotations themselves annotated with {@link Qualifier}.
+     *
+     * Resolvers should return static results that don't vary over time; therefore their results are cached.
+     * Adding a resolver will clear the local qualifier resolution cache, but is currently not guaranteed to have any effect on {@code SqlObjects}.
+     * Add any resolvers you may need <em>before</em> using any qualifiers-related features!
+     *
+     * @param resolver the resolver to be included in qualifier resolution
+     * @return this
+     */
+    public Qualifiers addResolver(Function<AnnotatedElement, Set<Annotation>> resolver) {
+        resolvers.add(resolver);
+        if (registry != null) {
+            QUALIFIER_CACHE.clear(registry);
+        }
+        return this;
+    }
+
     /**
      * Returns the set of qualifying annotations on the given elements.
+     *
      * @param elements the annotated elements. Null elements are ignored.
      * @return the set of qualifying annotations on the given elements.
      */
     public Set<Annotation> findFor(AnnotatedElement... elements) {
-        if (registry == null) {
-            return getQualifiers(elements);
-        }
-        return QUALIFIER_CACHE.get(elements, registry);
+        return registry == null
+            ? resolveQualifiers(elements)
+            : QUALIFIER_CACHE.get(elements, registry);
     }
 
-    private static Set<Annotation> getQualifiers(AnnotatedElement... elements) {
-        return Collections.unmodifiableSet(Arrays.stream(elements)
-                .filter(Objects::nonNull)
-                .map(AnnotatedElement::getAnnotations)
-                .flatMap(Arrays::stream)
-                .filter(anno -> anno.annotationType().isAnnotationPresent(Qualifier.class))
-                .collect(toSet()));
+    private Set<Annotation> resolveQualifiers(AnnotatedElement... elements) {
+        Set<AnnotatedElement> nonNullElements = Arrays.stream(elements)
+            .filter(Objects::nonNull)
+            .collect(toSet());
+
+        return resolvers.stream()
+            .flatMap(resolver -> nonNullElements.stream().map(resolver))
+            .flatMap(Collection::stream)
+            .collect(toSet());
     }
 
     @Override
     public Qualifiers createCopy() {
-        return this;
+        return new Qualifiers(this);
+    }
+
+    private static Set<Annotation> getQualifierAnnotations(AnnotatedElement element) {
+        return Arrays.stream(element.getAnnotations())
+            .filter(anno -> anno.annotationType().isAnnotationPresent(Qualifier.class))
+            .collect(toSet());
+    }
+
+    private static Set<Annotation> inspectQualifiedCarrier(AnnotatedElement element) {
+        return Arrays.stream(element.getAnnotations())
+            .filter(anno -> anno instanceof Qualified)
+            .map(Qualified.class::cast)
+            .map(Qualified::value)
+            .flatMap(Arrays::stream)
+            .map(AnnotationFactory::create)
```

Comment:
I worry that this approach does not work for any annotations that themselves have properties, as you can only specify their `Class`.

Reply:
Correct, but this is the most we can express within the `Qualified` meta-annotation approach. For annotations with attributes you'll have to register your own resolver.

Reply:
I guess I don't have a better suggestion at the moment anyway.